### DISCUSSION TITLE
trying to figure out runtime lists

### DISCUSSION THREAD

**Discussion by emaayan** [davetcc/tcMenu#335] [Q&A]
so i'm trying to create a timezone selection menu where one would selected a country then a city for timezone offset 
i would do the same for GPS coordinates
the problems is i can't figure out how to actually create the lists.. i mean assume i have an array of struct i've generated from a pre-made list how would i attach it to the menu 
i kept reading the docs, but they seem a little out of sync, like for example the initial raw field doesn't appear in the generator, i thought i would enter that number, and the callback woudl be used for each row to fill out it's name and value. 

![image](https://user-images.githubusercontent.com/13434544/232339977-f633535d-6264-49c9-a201-70a9284872e1.png)

**Comment by davetcc** [comment]
So when working with lists, there's a couple of steps, first, once the menu item is generated it will put a "renderFn" into your INO sketch. This will be called whenever anything happens on the list. You can read about that here - https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/menu-item-types/list-menu-item/

But in short, during setup or when the number of items changes, you set the number of items on the list, then as the list is drawn, it will call the render function to get the name and value for each item. The name appears on the left, the value on the right. The title is a special row, it is the row numbered `LIST_PARENT_ITEM_POS` and whatever text is returned from the render function for that row will be the text for the title and back button.

About the clearest example for how you intend to use it could be - https://github.com/davetcc/tcMenuLib/blob/master/examples/esp/esp32Amplifier/esp32Amplifier.ino#L187

**Reply by emaayan** [reply]
thanks, a little bit clearer now, but where do you place the set the number of items on list? (not sure what you mean during setup, cause i have no access to initialRows value in the generator

**Reply by davetcc** [reply]
You do it in your code, at whatever point you know how many rows you have. I'm assuming during setup or some point later you'd know. For example, if the list was called "MyList" and there were 10 items in the list, you would call.

    menuMyList.setNumberOfRows(10);

**Reply by emaayan** [reply]
you mean somewhere in my callback cause that's the only place i have access to myMenuList 


```
int CALLBACK_FUNCTION fnSetTimeZoneListRtCall(RuntimeMenuItem *item, uint8_t row, RenderFnMode mode, char *buffer, int bufferSize)
{
    const bool isTitle = row == LIST_PARENT_ITEM_POS;
    switch (mode)
    {
    case RENDERFN_INVOKE:
        // item->setNumberOfRows(10);
        //  TODO - your code to invoke goes here - row is the index of the item
        if (isTitle){
            item->setNumberOfRows(10);
        }
        
        if (!isTitle && renderer.getDialog() && !renderer.getDialog()->isInUse())
        {
         
            renderer.getDialog()->setButtons(BTNTYPE_NONE, BTNTYPE_CLOSE);
            renderer.getDialog()->showRam("List select", false);
            char sz[10];
            ltoaClrBuff(sz, row, 3, NOT_PADDED, sizeof sz);
            renderer.getDialog()->copyIntoBuffer(sz);
        }
        return true;
    case RENDERFN_NAME:
        if (isTitle)
        {
            strcpy(buffer, "TimeZones");
        }
        else
        {
            strncpy(buffer, "ISR", bufferSize);
            ltoaClrBuff(buffer, row, 3, NOT_PADDED, bufferSize);
        }
        return true;
    case RENDERFN_VALUE:
        if (isTitle)
        {
            strcpy(buffer, ">>");
        }
        else
        {
            strncpy(buffer, "Israel", bufferSize);
            fastltoa(buffer, row, 3, NOT_PADDED, bufferSize);
        }
        return true;
    case RENDERFN_EEPROM_POS:
        return 0xffff; // lists are generally not saved to EEPROM
    default:
        // return textItemRenderFn(item, row, mode, buffer, bufferSize); // pass thru to default
        return false;
    }
```

**Comment by davetcc** [comment]
No, you always have access to all menu items, take a look at the file  yourProjectName_menu.h in your project, it exports every menu item as a global. You can use them anywhere.

For example a snippet from picoAdafruitDashboard_menu.h - 

    extern BooleanMenuItem menuSettingsCheckBox;
    extern BooleanMenuItem menuSettingsYesNo;
    extern Rgb32MenuItem menuSettingsRGB;
    ... more definitions

**Reply by emaayan** [reply]
oooh that was the missing link..
and only when they select a menuItem from that list i would need to generate another list based on their entry

**Reply by emaayan** [reply]
> No, you always have access to all menu items, take a look at the file yourProjectName_menu.h in your project, it exports every menu item as a global. You can use them anywhere.
> 
> For example a snippet from picoAdafruitDashboard_menu.h -
> 
> ```
> extern BooleanMenuItem menuSettingsCheckBox;
> extern BooleanMenuItem menuSettingsYesNo;
> extern Rgb32MenuItem menuSettingsRGB;
> ... more definitions
> ```

so now i just realized all this is only good for only the first level menu 
if i want to create a 2nd level menu that would be created in run-time i can't do that. 
i know there's a insertSubmenu in the dialogBox 
but from what i saw in dynamicMenuIItme those menu are actually defined in code statically with 
```
RENDERING_CALLBACK_NAME_INVOKE(fnExtraDialogLine, textItemRenderFn, "Detail: ", -1, NO_CALLBACK)
TextMenuItem secondItem(fnExtraDialogLine, nextRandomId(), 12, nullptr);
```
so assuming i would like to create a submenu based solely on a list i've created from a previous menu that would be an issue. 
so i thought about doing something like this.

```
int CALLBACK_FUNCTION fnsetTimeZoneRtCall(RuntimeMenuItem *item, uint8_t row, RenderFnMode mode, char *buffer, int bufferSize)
{
    const bool isTitle = row == LIST_PARENT_ITEM_POS;
    switch (mode)
    {
    case RENDERFN_INVOKE:
    {

        BaseDialog *dlg = renderer.getDialog();
        if (!isTitle && dlg && !dlg->isInUse() && dlg->isMenuItemBased())
        {
            country ctr = getCountry(row);
            // city cities[ctr.totalCities];
            city cities[] = {
                {"IL", 0, "ehlhana", 232.22, 32.0, 3.0, "AsiaEurp", "sda32"}
                , {"US", 1, "dsdlhana", 244.22, 344.0, 3.0, "AsiaEurpsd", "sda32232"}};
            // getCities(ctr, cities);

            MenuBasedDialog *m_dlg = static_cast<MenuBasedDialog *>(dlg);
            m_dlg->setButtons(BTNTYPE_NONE, BTNTYPE_CLOSE);
            for (size_t i = 0; i < sizeof(cities) / sizeof(cities[0]); i++)
            {
                city ct= cities[i];
                m_dlg->setUserData((void *)cities);
                ListRuntimeMenuItem cityItem = ListRuntimeMenuItem(nextRandomId(), 0, fnsetCityRtCall);
                
                m_dlg->insertMenuItem(&cityItem);
            }

            m_dlg->showRam(ctr.name, false);
        }
        return true;
    }
    case RENDERFN_NAME:
    {
        if (isTitle)
        {
            strncpy(buffer, "TimeZones", bufferSize);
        }
        else
        {

            const country c = getCountry(row);
            strncpy(buffer, c.code, bufferSize);
            // strncpy(buffer, "ISR", bufferSize);
            // ltoaClrBuff(buffer, row, 3, NOT_PADDED, bufferSize);
        }
        return true;
    }
    case RENDERFN_VALUE:
    {
        if (isTitle)
        {
            strncpy(buffer, ">>", bufferSize);
        }
        else
        {

            //   strncpy(buffer, "Israel", bufferSize);
            // fastltoa(buffer, row, 3, NOT_PADDED, bufferSize);
            // ExtendedZone zone = manager.getZoneForIndex(row);
            // ace_common::PrintStr<64> printer;
            // zone.printShortNameTo(printer);
            // const char * n=printer.cstr();

            // strncpy(buffer,n, bufferSize);
            const country c = getCountry(row);
            strncpy(buffer, c.name, bufferSize);
            // fastltoa(buffer, row, 3, NOT_PADDED, bufferSize);
        }
        return true;
    }
    case RENDERFN_EEPROM_POS:
        return 0xffff; // lists are generally not saved to EEPROM
    default:
        // return textItemRenderFn(item, row, mode, buffer, bufferSize); // pass thru to default
        return false;
    }
}

```

**Reply by davetcc** [reply]
On this front, you're a bit ahead of the game! The code has been there for a long time to do it, but has not been used that much as far as I know. Maybe it has and others please do correct me there if you've used it.

You could either do that, or you can create a menu entirely at runtime and add it into the general tree structure, but this is only good if you want to be able to navigate to the structure in the regular way (IE from the menu). To do that you can use the menuMgr helper function:

	/**
	 * Adds a menu item into the tree directly after the existing item provided. Never add an item that's already in
	 * the tree. Don't forget that if you use silent, to call menuStructureChanged() after you're done adding items.
	 * @param existing where in the tree the new item is to be added.
	 * @param toAdd the item that should be added, must not be in the tree already.
	 * @param silent do not run the structure changed callback.
	 */
	void addMenuAfter(MenuItem* existing, MenuItem* toAdd, bool silent = false);

If you don't really want the dialog support, just a temporary menu, you can just create an enitre structure as described in the menu type docs, and pass that to `navigateToMenu(menuNewStructureRootItemPtr, menuItemToActivatePtr, true)` to display it without affecting the history (last param true means skip putting into history).

One thing that I think is missing, and I will add it into 3.2 as it has been annoying me for a while, is to be able to define runtime menu items with menu info blocks, it has already been done for submenu items and it makes creating them at runtime using new far easier. This is just a case of moving a little bit of code around. Then instead of having to create the RtFn stuff, in the simplest case, you'd just do something like:

    auto myItem = new TextMenuItem(new AnyMenuInfo("My Name", myId, myEeeprom, myCallbackFn), INFO_LOCATION_RAM, size, next);

It has a very small memory requirement, but what I've noticed is that even on Uno we are not generally memory constrained, it is normally Flash that is the issue on smaller boards, the libraries are quite efficient, and unfortunately, we cannot continue to drive everything around it running on an Uno. If need be, we'll create a compile flag to handle that case.

**Reply by emaayan** [reply]
but the problem is that the only menu types that i can add at runtime are ListRuntimeMenuItem and those require a callback function of their own to dispaly the name and value.

**Reply by davetcc** [reply]
You can add any you like at runtime, please read through the dynamicMenuItems example and notice that the Oven menu is added completely at runtime, from scratch. But as I said, you're ahead of the curve here, this support has not been used that much, as is still to some extent in development. To get more or less what I think that you want, you may have to wait until 3.2 to come out.

I think most users now need this simplification to make creating runtime items easier, and there will not be a big release after 3.2 for a while, so the support needs to go in there, as it needs a lot of testing before release. A simple change but needs heavy testing.

**Reply by davetcc** [reply]
This will get easier from 3.2, but it is entirely possible, although a bit messy in 3.1 - https://github.com/davetcc/tcMenuLib/issues/197

**Reply by emaayan** [reply]
so untill 3.2 arrives i think i would with compromise of 2 menus 
several problems, 
i need to exit the countries menu as soon as the user selects a country 
when they select a country i need to save to eeprom as well as prep the json for the cities 
it would be nice if i had a callback for a runtime menu to be called BEFORE the user enters the menu to prep the list
 
![image](https://user-images.githubusercontent.com/13434544/234696664-167c4843-6a3a-4e16-be26-0b2c22d9a2c6.png)

**Reply by davetcc** [reply]
> it would be nice if i had a callback for a runtime menu to be called BEFORE the user enters the menu to prep the list

See menu observer: https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/menumanager-and-iteration/#listening-for-changes-on-menu-manager

**Reply by emaayan** [reply]
thanks i'll look into it.. 

how do i get the menu to go back after someone click on a runtimeList? so far i've only managed to do this:

but as you can see the selector is actually on 2 places 
```
case RENDERFN_INVOKE:
    {
        const JsonObject country = countries[row];
        char const *countryCode = country["countryCode"];
        const size_t totalCities = loadCity(countryCode);
        menuCity.setNumberOfRows(totalCities);

        menuMgr.resetMenu(false);
        menuCountry.setActive(true);
        return true;
    }

```

![image](https://user-images.githubusercontent.com/13434544/235128389-f7b52d6c-66f6-4a1c-9957-ba4c6dce2750.png)

**Reply by davetcc** [reply]
To change the active item use `menuMgr.changeMenu(menuItemToBeActive);`, this makes sure any editing is ended and that the state is reset properly first.

Or you can navigate and do that at the same time

`menuMgr.navigateToMenu(rootItemInMenuPtr, itemToActivatePtr, boolShouldGoInNavHistiory);`

See both of the above in the menu manager documentation.

**Reply by emaayan** [reply]
thanks, i tried almsot eveyr combo but only this worked by having the arrow pointed at menuCountry 
        menuMgr.resetMenu(false);
        menuMgr.changeMenu(&menuCountry);

doing this didn't work (meaning it didn't set the arrow on the country entry) 
        //  menuMgr.navigateToMenu(&menuLocation,&menuCountry);

**Reply by emaayan** [reply]
> To change the active item use `menuMgr.changeMenu(menuItemToBeActive);`, this makes sure any editing is ended and that the state is reset properly first.
> 
> Or you can navigate and do that at the same time
> 
> `menuMgr.navigateToMenu(rootItemInMenuPtr, itemToActivatePtr, boolShouldGoInNavHistiory);`
> 
> See both of the above in the menu manager documentation.

is there a built-in way i can save to eeprom (i think tcmenu has has setting abstraction for saving stuff) what the user have selected during invoke?

**Reply by davetcc** [reply]
If it is menu state, you can either use the built in tcMenu save and load functions if they work for you - https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/menu-eeprom-integrations/

There are a few possible strategies on how to save discussed on that page, but never call `save` during the callback invoke method, you'll destroy the ROM in days, they are measured in thousands of cycles.

Given that, once you've configured a ROM in code generator, it would be as easy as `menuMgr.save()` in the right place.

If you wanted to save custom data, you could use the helper that is available from `menuMgr.getEepromAbstraction()`, it is a pointer to the EEPROM you configured from code generator - https://www.thecoderscorner.com/products/arduino-libraries/io-abstraction/eeprom-impl-seamless-8-and-32-bit/

**Reply by emaayan** [reply]
can i use the commit callback to save to eepROm?
i was thingking somethign like this: 
this will also fill the subMenues and call eppRom classes
(shame i don't see any generated const numbers for menus)

```
#include <ArduinoEEPROMAbstraction.h>
#include <EEPROM.h>
const unsigned int romStart = 400;
ArduinoEEPROMAbstraction eepromWrapper(&EEPROM);
void onCommit(int menuId)
{

    switch (menuId)
    {
    case 5: // case Country
//eepromWrapper.write16()
//EEPROM.commit();
        break;
    case 6: // city

        break;
        
    default:
        break;
    }
}
void initSystem()
{
    status status = loadCountries();
    if (status.code > 0)
    {
        setupMenu();
        menuCountry.setNumberOfRows(countries.as<JsonArray>().size());
        menuMgr.setItemCommittedHook(onCommit);
        //  renderer.takeOverDisplay(myDisplayCallback);
    }
    else
    {
        writeError("Error %s", status.error);
    }
}
```

**Reply by davetcc** [reply]
You should factor in about 10000-100000 cycles on an EEPROM depending on type, so you could call that commit around that many times before running the risk of starting to see failures. Personally, I wouldn't do that, there are discussions around ways to proceed in the link I provided above, including detecting power loss and saving when power is lost.

**Reply by davetcc** [reply]
End of the day it depends how often you intend to update that item.

For me the best solution is https://www.thecoderscorner.com/electronics/microcontrollers/psu-control/detecting-power-loss-in-powersupply/

The second best is to have a timer that checks if anything needs saving every N minutes and then does the save.

The third best is a save menu option on the menu (and I've often for simple stuff just gone with this as it is so simple).

**Reply by emaayan** [reply]
> End of the day it depends how often you intend to update that item.
> 
> For me the best solution is https://www.thecoderscorner.com/electronics/microcontrollers/psu-control/detecting-power-loss-in-powersupply/
> 
> The second best is to have a timer that checks if anything needs saving every N minutes and then does the save.
> 
> The third best is a save menu option on the menu (and I've often for simple stuff just gone with this as it is so simple).

but the thing is , i don't plan on this menu being called often, infact i plan on it being called maybe only once during the device operation, because the entire menu shouldn't be active in the first.. the primary functon of the device is to display a clock and hebrew festivals dates which require the user to input timezones and location, so i figured i could have both in one shot in a config menu that will be saved once they select the proper location and be done with it. 
that's why i figured the commit would be a good candidate.

**Reply by davetcc** [reply]
Agreed, if it has a low rate of change and you're using the commit callback, which only fires when the editing ends, then it should be quite safe as long as it's not edited many times a day or something.

**Reply by emaayan** [reply]
the problem is that neither onCommit seems to getCalled on any menu change 

and all the methods except for menuEditEnded for MyObserver 
they also don't seem to get called for any of my runtime addedMenu 
and my original intention was to have menduEditEnded get Called on a runtimeMenuItemSelection 


```
void onCommit(int menuId)
{

    if (menuCountry.getId() == menuId)
    {
        // Serial.print(menuId);
        const char *ctr = saveCountry();
        loadMenuCity(ctr);
    }
    if (menuCity.getId() == menuId)
    {
    }
}

class MyObserver : public MenuManagerObserver
{
public:
    void structureHasChanged() override
    {
    }

    bool menuEditStarting(MenuItem *item) override
    {

        return true; // allow editing to start
    }

    void menuEditEnded(MenuItem *item) override
    {
        menuid_t menuId = item->getId();
        if (menuCountry.getId() == menuId)
        {
            // Serial.print(menuId);
            const char *ctr = saveCountry();
            loadMenuCity(ctr);
        }
        if (menuCity.getId() == menuId)
        {
        }
    }

    void activeItemHasChanged(MenuItem *newActive) override
    {
        menuid_t menuId = newActive->getId();
        if (menuCountry.getId() == menuId)
        {
        }
    }

} myObserver;

void initSystem()
{

    status status = loadCountries();
    if (status.code > 0)
    {

        u_int32_t initConf = loadInit();
        loadCountry();
        setupMenu();

       
        const size_t totalCountries = countries.as<JsonArray>().size();
        menuCountry.setNumberOfRows(totalCountries);
         menuMgr.addChangeNotification(&myObserver);
      //  menuMgr.setItemCommittedHook(&onCommit);
        //  renderer.takeOverDisplay(myDisplayCallback);
    }
    else
    {
        writeError("Error %s", status.error);
    }
}

void setup()
{
    Serial.begin(115200);
    initDisplay();

    if (!LITTLEFS.begin(false))
    {
        writeError("%s", "LITTLEFS Mount Failed");
        return;
    }
    else
    {
        initSystem();
    }
}
```

**Reply by davetcc** [reply]
please raise another discussion, think about others trying to search for solutions later, there are several problems already in here.

**Comment by emaayan** [comment]
it this relevant for runtimeMenuItem as well?

**Reply by davetcc** [reply]
Anything that has an EEPROM location in the designer will save to EEPROM if the location is not -1.
